File: 78-subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        arr = []

        def backtrack(arr, comb, i):
            if i == len(nums):
                arr.append(comb[:])
                return
            if len(comb) == len(nums):
                arr.append(comb[:])
                return

            comb.append(nums[i])
            backtrack(arr, comb, i + 1)
            comb.pop()
            backtrack(arr, comb, i + 1)
        backtrack(arr, [], 0)

        return arr

File: 78-subsets/subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const arr = []
    
    const dfs = (i, list) => {
        if (i === nums.length) {    
            arr.push([...list])
        return
        }

        // creates decision tree where we make 2 distinct choices
        dfs(i + 1, list)
        dfs(i + 1, [...list, nums[i]])
    }
    dfs(0, [])

    return arr
};